src/apexguru/apex-guru-service.ts

```diff
@@ -29,22 +33,67 @@ export async function isApexGuruEnabledInOrg(outputChannel: vscode.LogOutputChan
 	}
 }
 
-export async function runApexGuruOnFile(selection: vscode.Uri, outputChannel: vscode.LogOutputChannel) {
+export async function runApexGuruOnFile(selection: vscode.Uri, runInfo: RunInfo) {
+	const {
+		diagnosticCollection,
+		commandName,
+		outputChannel
+	} = runInfo;
+	const startTime = Date.now();
 	try {
-        const requestId = await initiateApexGuruRequest(selection, outputChannel);
-		// TODO: Logging the request Id for easy QA. Future stories will use this requestId to poll and retrieve the Apex Guru report.
-        outputChannel.appendLine('***Apex Guru request Id:***' + requestId);
+		return await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification
+		}, async (progress) => {
+			progress.report(messages.apexGuru.progress);
+			const connection = await CoreExtensionService.getConnection();
+			const requestId = await initiateApexGuruRequest(selection, outputChannel, connection);
+			outputChannel.appendLine('***Apex Guru request Id:***' + requestId);
+
+			const queryResponse: ApexGuruQueryResponse = await pollAndGetApexGuruResponse(connection, requestId);
+
+			const decodedReport = Buffer.from(queryResponse.report, 'base64').toString('utf8');
+			outputChannel.appendLine('***Retrieved analysis report from ApexGuru***');
+
+			const ruleResult = transformStringToRuleResult(selection.fsPath, decodedReport);
+			new DiagnosticManager().displayDiagnostics([selection.fsPath], [ruleResult], diagnosticCollection);
+			TelemetryService.sendCommandEvent(Constants.TELEM_SUCCESSFUL_APEX_GURU_FILE_ANALYSIS, {
+				executedCommand: commandName,
+				duration: (Date.now() - startTime).toString()
+			});
+		});
     } catch (e) {
         const errMsg = e instanceof Error ? e.message : e as string;
         outputChannel.appendLine('***Apex Guru initiate request failed***');
         outputChannel.appendLine(errMsg);
     }
 }
 
-export async function initiateApexGuruRequest(selection: vscode.Uri, outputChannel: vscode.LogOutputChannel): Promise<string> {
+export async function pollAndGetApexGuruResponse(connection: Connection, requestId: string) {
+	let queryResponse: ApexGuruQueryResponse;
+	const retryInterval = Constants.APEX_GURU_RETRY_INTERVAL;
+	let retryCount = Constants.APEX_GURU_RETRY_COUNT;
+
+	while (retryCount > 0) {
+		queryResponse = await connection.request({
```

Comment:
Does the request method allow us to send in a connection timeout in case the server freezes? Ideally the individual request timeout should always be close to totalTimeout - currentDuration

Reply:
I don't think so. I think this could be one of the requirements we could ask from the Apex Guru team.